presto-docs/src/main/sphinx/connector/kudu.rst

```diff
@@ -0,0 +1,494 @@
+==============
+Kudu Connector
+==============
+
+The kudu connector allows querying, inserting and deleting data in `Apache Kudu`_
+
+.. _Apache Kudu: https://kudu.apache.org/
+
+
+.. contents::
+    :local:
+    :backlinks: none
+    :depth: 1
+
+
+Compatibility
+-------------
+
+Connector is compatible with all Apache Kudu versions starting from 1.0.
+
+If the connector uses features that are not available on the target server, an error will be returned.
+Apache Kudu 1.7.0 is currently used for testing.
+
+
+Configuration
+-------------
+
+To configure the Kudu connector, create a catalog properties file
+``etc/catalog/kudu.properties`` with the following contents,
+replacing the properties as appropriate:
+
+  .. code-block:: none
+
+       connector.name=kudu
+
+       ## List of Kudu master addresses, at least one is needed (comma separated)
+       ## Supported formats: example.com, example.com:7051, 192.0.2.1, 192.0.2.1:7051,
+       ##                    [2001:db8::1], [2001:db8::1]:7051, 2001:db8::1
+       kudu.client.master-addresses=localhost
+
+       ## Optional restriction of tablets for specific tenant.
+       ## If a tenant is set, only Kudu tablets starting with `<tenant>.` will
+       ## be visible in Presto
+       #kudu.session.tenant=mytenant
+
+       #######################
+       ### Advanced Kudu Java client configuration
+       #######################
+
+       ## Default timeout used for administrative operations (e.g. createTable, deleteTable, etc.)
+       #kudu.client.defaultAdminOperationTimeout = 30s
+
+       ## Default timeout used for user operations
+       #kudu.client.defaultOperationTimeout = 30s
+
+       ## Default timeout to use when waiting on data from a socket
+       #kudu.client.defaultSocketReadTimeout = 10s
+
+       ## Disable Kudu client's collection of statistics.
+       #kudu.client.disableStatistics = false
+
+
+Querying Data
+-------------
+
+A Kudu table named ``mytable`` is available in Presto as table
+``kudu.default.mytable``. A Kudu table containing a dot is considered as
+a schema/table combination, e.g. ``dev.mytable`` is mapped to the Presto
+table \`kudu.dev.mytable. Only Kudu table names in lower case are
+currently supported.
+
+Before using any tablets, it is needed to create the default schema,
+e.g.
+
+  .. code:: sql
+
+    CREATE SCHEMA default;
+
+Example
+~~~~~~~
+
+-  Create default schema if needed:
+
+  .. code:: sql
+
+    CREATE SCHEMA IF NOT EXISTS default;
+
+-  Now you can use any Kudu table, if it is lower case and contains no
+   dots.
+-  Alternatively you can create a users table with
+
+  .. code:: sql
+
+    CREATE TABLE users (
+      user_id int,
+      first_name varchar,
+      last_name varchar
+    ) WITH (
+     column_design = '{"user_id": {"key": true}}',
```

Comment:
I do not like passing json as configuration. Is there a way to use complex types? Like in hive:
```
CREATE TABLE hive.web.page_views (
...
)
WITH (
  partitioned_by = ARRAY['ds', 'country'],
..
)
```
I in your case it would be a nested map instead of array. Or maybe you could more specific properties like `column_design_user_id_key = true`.

Or maybe something like:
```
CREATE TABLE users (
     user_id int,

      first_name varchar
  ) WITH (
    column_keys = ARRAY ['user_id'], -- or key_columns
    compression = MAP(ARRAY['user_id', 'first_name'], ARRAY['user_id_compression', 'first_name_compression'])
);
```
@electrum do you have any ideas about the above syntax?

Reply:
I would use the complex type. It seems better from a type safety perspective, and avoids the issues of escaping JSON in a string.

I agree the current way of creating a map using two arrays is ugly. We’ve discussed having better syntax for it. At that point, this will automatically take advantage of it and be better than JSON.

Reply:
I don't understand, how I can use the complex type here.
For example if you have a complicated partition design like
```sql
partition_design = '{
  "hash":[{"columns":["serialno", "col2"], "buckets": 4}],
  "range": {"columns":["event_time", "col3"]}
}'
```
with complex types this would translate into
```sql
MAP(
 ARRAY['hash',ARRAY[MAP(ARRAY['columns',ARRAY['serialno','col2']],ARRAY['buckets',4])]],
 ARRAY['range',MAP(ARRAY['columns',ARRAY['event_time','col3']])]
)
```
But this is not allowed in Presto, as `All ARRAY elements must be the same type`

Reply:
What do you think about removing nested properties?

For example:
```
partition_design = '{
  "hash":[{"columns":["serialno", "col2"], "buckets": 4}],
  "range": {"columns":["event_time", "col3"]}
}'
```
could represented as:
```
partition_design_hash_columns = ARRAY["serialno", "col2"],
partition_design_hash_buckets = 4,
partition_design_range_columns = ARRAY ["event_time", "col3"]
```
Or even:
```
hash_columns = ARRAY["serialno", "col2"],
hash_buckets = 4,
range_columns = ARRAY ["event_time", "col3"]
```

That way it looks simpler. @MartinWeindel What do you think about it?

Reply:
Well, for simple cases it looks good.
There are several complications:
1. You can have multiple hash partitioning groups in Kudu. It's probably a exotic use case, but it should be possible to use it.
Following this idea, it could be represented as:
```
partitioning_hash_columns = ARRAY['serialno', 'col2'],
partitioning_hash_buckets = 4,
partitioning_hash2_columns = ARRAY['col3'],
partitioning_hash2_buckets = 10,
partitioning_range_columns = ARRAY['event_time'],
```

2. For the column design it still looks even more awkward with structural types than with JSON:
- JSON:
  ```
  column_design = '{"serialno": {"key": true}, 
    "column1": {"key": true, "encoding": "dictionary", "compression": "LZ4"}, 
    "column2": {"nullable": true, "encoding": "dictionary"}}',
  ```
- Structural types:
  ```
  column_design = MAP(ARRAY['serialno', 'column1','column2'], ARRAY[
    MAP(ARRAY['key'], ARRAY['true']),  
    MAP(ARRAY['key','encoding','compression'],ARRAY['true','dictionary','compression']), 
    MAP(ARRAY['nullable','encoding'],ARRAY['true','dictionary'])]) 
  ```

3. And for range partitions I have still no clue at all how to support all variants.
See docu [range_partitions](https://github.com/MartinWeindel/presto/blob/master/presto-docs/src/main/sphinx/connector/kudu.rst#table-property-range_partitions) for details.

So creating Kudu tables with Presto will never feel like "first class support" with the current SQL parser possibilities.
The Impala SQL parser has support for Kudu column properties, primary key and Kudu partitioning design and range partition values, as you can see [here](https://www.cloudera.com/documentation/enterprise/5-12-x/topics/impala_create_table.html)

Much effort for very little outcome...

Reply:
For the column design another idea is to extend the Presto SQL paser and allow column properties similar to table properties, e.g.
```sql
CREATE TABLE kudu.default.sample (
  serialno varchar,
  col1 varchar WITH (encoding = 'dictionary', compression = 'LZ4'),  
  col2 varchar WITH (nullable = 'true', encoding = 'prefix)
) WITH (
  primary_key = ARRAY['serialno', 'column1'],...
);
```

This would allow to improve `ALTER TABLE ADD COLUMN`, too:
```sql
ALTER TABLE kudu.default.sample ADD COLUMN
  column3 bigint WITH (nullable = 'true', encoding='rle'); 
```

Reply:
It looks like ANSI SQL does not define anything like column properties.

```
<column definition> ::=
<column name> [ <data type or domain name> ]
[ <default clause> | <identity column specification> | <generation clause> ]
[ <column constraint definition>... ]
[ <collate clause> ]
<data type or domain name> ::=
<data type>
| <domain name>
<column constraint definition> ::=
[ <constraint name definition> ] <column constraint> [ <constraint characteristics> ]
<column constraint> ::=
NOT NULL
| <unique specification>
| <references specification>
| <check constraint definition>
<identity column specification> ::=
GENERATED { ALWAYS | BY DEFAULT } AS IDENTITY
[ <left paren> <common sequence generator options> <right paren> ]
<generation clause> ::= <generation rule> AS <generation expression>
<generation rule> ::= GENERATED ALWAYS
<generation expression> ::= <left paren> <value expression> <right paren>
```
@martint What do you think about such syntax for column properties?

However I like this approach, very much. It should not be much hard to add it to parser. The only problem with it is that it does not solve hash or range partitioning configuration. Do you (@MartinWeindel) have any idea how to solve this?

Reply:
For partition design, your idea to flatten to nested properties looks good and can be extended to cover all cases. I changed the wording to partition_by_...`.  E.g.:
```sql
partition_by_hash_columns = ARRAY['serialno', 'col2'],
partition_by_hash_buckets = 4,
partition_by_range_columns = ARRAY['event_time'],
```
If there are multiple hash partition groups, then the value for `partition_by_hash_columns` has to be an array of arrays, the value for `partition_by_hash_buckets` an ARRAY of ints, e.g.
```sql
partition_by_hash_columns = ARRAY[ARRAY['serialno', 'col2'], ARRAY['col3']]
partition_by_hash_buckets = ARRAY[4, 10],
```

But for defining range partitions, I suggest to stay with the JSON expression to keep it similar to the 
`add_range_partition` call like
```sql
CALL kudu.system.add_range_partition('myschema', 'events', '{"lower": "2018-01-01", "upper": "2018-06-01"}')  
```
As example
```sql
CREATE TABLE sample (...) WITH (
   partition_by_range_columns = ARRAY['event_time'],
   range_partitions = ARRAY[
     '{"lower": null, "upper": "2018-06-01"}',
     '{"lower": "2018-01-01", "upper": "2018-06-01"}']
);
```
Replacing JSON here is hard, because the lower and upper bounds types are depending on the column type.
 Examples:
  
  | Presto Data Type | JSON string example |
  | ---------------- | ------------------- |
  | BIGINT           | '{"lower": 0, "upper": 1000000}' | 
  | SMALLINT         | '{"lower": 10, "upper": null}' | 
  | VARCHAR          | '{"lower": "A", "upper": "M"}' | 
  | TIMESTAMP        | '{"lower": "2018-02-01T00:00:00.000", "upper": "2018-02-01T12:00:00.000"}' | 
  | BOOLEAN          | '{"lower": false, "upper": true}' | 
  | VARBINARY        | values encoded as base64 strings |
  
Unbounded bound must be possible, too. (The JSON value `null` is used in this syntax).
I also remember to have trouble with `CALL` when I tried to use `ARRAY` instead of `STRING`  arguments.

Reply:
I like the idea of column properties. Table (and schema) properties are an extension to the specification, so I don't see a problem adding them for columns. We simply haven't needed them yet.

Regarding array arguments for a procedure, here's a working example: https://gist.github.com/electrum/c85f6c4b3a45e0270f868e818dae49dc

I designed `Argument` to take a string or `TypeSignature` rather than a `Type` to avoid the need to lookup complex types in `TypeRegistry` (just let the engine do it).

Reply:
@electrum I still have trouble with using typed arguments instead of a simple JSON string.
The problem is that I need dynamic types. I.e. the concrete type of the range bound values are depending on the column types.
For example, if the range partition is a single string column, then I need the type `ARRAY(MAP(ARRAY(varchar),ARRAY(varchar)))`. If it is a single bigint column, I need `ARRAY(MAP(ARRAY(varchar),ARRAY(bigint)))`. If the range partition consists of multiple columns with different types, you would have to switch to the ROW type, e.g. `ARRAY(MAP(ARRAY(varchar), ARRAY(ROW(varchar, bigint))))`, etc.
With JSON it is quite easy to handle all the cases dynamically:
- single range partition column `varchar`: `'{"lower": "A", "upper": "M"}'` 
- two range partition columns `varchar,bigint`: `'{"lower": ["A", null], "upper": ["A", 10000]}'` 

Any ideas?